<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Gubbins : Genealogies Unbiased By recomBinations In Nucleotide Sequences">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Gubbins - Genealogies Unbiased By recomBinations In Nucleotide Sequences</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/nickjcroucher/gubbins">View source code on GitHub</a>

          <h1 id="project_title">Gubbins</h1>
          <h2 id="project_tagline">Genealogies Unbiased By recomBinations In Nucleotide Sequences</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/nickjcroucher/gubbins/zipball/master">Download source code as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/nickjcroucher/gubbins/tarball/master">Download source code as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h2>
<a name="overview" class="anchor" href="#overview"><span class="octicon octicon-link"></span></a>Overview</h2>

<p>Since the introduction of high-throughput, second-generation DNA sequencing technologies, there has been an enormous increase in the size of datasets being used for estimating bacterial population phylodynamics. Although many phylogenetic techniques are scalable to hundreds of bacterial genomes, methods which have been used for mitigating the effect of mechanisms of horizontal sequence transfer on phylogenetic reconstructions cannot cope with these new datasets. Gubbins (Genealogies Unbiased By recomBinations In Nucleotide Sequences) is an algorithm that iteratively identifies loci containing elevated densities of base substitutions while concurrently constructing a phylogeny based on the putative point mutations outside of these regions. Simulations demonstrate the algorithm generates highly accurate reconstructions under realistic models of short-term bacterial evolution, and can be run in only a few hours on alignments of hundreds of bacterial genome sequences.</p>

<h3>The paper</h3>
The paper is available from Nucleic Acids Research (open access):
<a href="http://nar.oxfordjournals.org/content/43/3/e15">Rapid phylogenetic analysis of large samples of recombinant bacterial whole genome sequences using Gubbins</a>.

<p><em>Croucher N. J., Page A. J., Connor T. R., Delaney A. J., Keane J. A., Bentley S. D., Parkhill J., Harris S.R.</em> "Rapid phylogenetic analysis of large samples of recombinant bacterial whole genome sequences using Gubbins". doi:10.1093/nar/gku1196, Nucleic Acids Research, 2014.
</p>

<h3>
<a name="downloads" class="anchor" href="#downloads"><span class="octicon octicon-link"></span></a>Installation</h3>
<p>Detailed installation instructions are available in the <a href="https://github.com/nickjcroucher/gubbins/blob/master/README.md">README file</a>.
    If you already have anaconda installed, then run:<br>
        <code class=shell>
        conda create -n gubbins<br>
        conda activate gubbins<br>
        conda config --add channels r<br>
        conda config --add channels defaults<br>
        conda config --add channels conda-forge<br>
        conda config --add channels bioconda<br>
        conda install gubbins
        </code>
</p>

<h3>User manual</h3>
Instructions on how to use the software are detailed in the
    <a href="https://github.com/nickjcroucher/gubbins/blob/master/docs/gubbins_manual.md"> manual</a><br>

<h3>Quickstart tutorial with example datasets</h3>
<ul>
  <li>A small set of <em>S. pneumoniae</em> PMEN3 assemblies can be downloaded <a href="https://figshare.com/ndownloader/files/33486593">here</a> (6.5 Mb archive of assemblies and a reference annotation). A brief tutorial on how to analyse such data is here <a href="https://github.com/nickjcroucher/gubbins/blob/master/docs/gubbins_tutorial.md">here</a></li>
  <li>A small <em>S. pneumoniae</em> PMEN1 dataset can be downloaded <a href="https://figshare.com/ndownloader/files/33471200">here</a> (6.4 Mb multi-FASTA whole genome alignment file). The expected output is <a href="https://github.com/nickjcroucher/gubbins/tree/master/example_data/PMEN1">here</a></li>
  <li>A small <em>S. aureus</em> ST239 dataset can be downloaded <a href="https://figshare.com/ndownloader/files/33472622">here</a> (11.3 Mb multi-FASTA whole genome alignment file). The expected output is <a href="https://github.com/nickjcroucher/gubbins/tree/master/example_data/ST239">here</a></li></li>
</ul>

<h3>Feedback/Issues</h3>
<p>Please report problems to the <a href="https://github.com/nickjcroucher/gubbins/issues">issues page</a>.</p>

<h3>Frequently asked questions</h3>
<h4>What type of dataset can be analysed with Gubbins?</h4>
<p>Gubbins detects recombinations through the locally-elevated densities of polymorphisms that arise when segments of sequence are acquired from a donor that is genetically divergent from the set of sequences being analysed. We recommend that you divide your population into strains using <a href="https://poppunk.net/">PopPUNK</a>, and use Gubbins on a whole genome alignment of each strain separately.</p>
<h4>Can the output of Roary be used as the input to Gubbins?</h4>
<p>The output of <a href="http://sanger-pathogens.github.io/Roary/">Roary (the pan genome pipeline)</a> cannot be used as the input to Gubbins.
Gubbins requires a whole genome alignment as input, in order to analyse the spatial distribution of base substitutions.</p>
<h4>Can an alignment of polymorphic sites be used as the input to Gubbins?</h4>
<p>An alignment of polymorphic sites cannot be used as the input to Gubbins.
Gubbins requires a whole genome alignment as input, in order to analyse the spatial distribution of base substitutions.</p>

      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Gubbins maintained by <a href="https://github.com/nickjcroucher">Nick Croucher</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
