<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Gubbins : Genealogies Unbiased By recomBinations In Nucleotide Sequences">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Gubbins</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/sanger-pathogens/gubbins">View source code on GitHub</a>

          <h1 id="project_title">Gubbins</h1>
          <h2 id="project_tagline">Genealogies Unbiased By recomBinations In Nucleotide Sequences</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/sanger-pathogens/gubbins/zipball/master">Download source code as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/sanger-pathogens/gubbins/tarball/master">Download source code as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a name="overview" class="anchor" href="#overview"><span class="octicon octicon-link"></span></a>Overview</h1>

<p>Since the introduction of high-throughput, second-generation DNA sequencing technologies, there has been an enormous increase in the size of datasets being used for estimating bacterial population phylodynamics. Although many phylogenetic techniques are scalable to hundreds of bacterial genomes, methods which have been used for mitigating the effect of mechanisms of horizontal sequence transfer on phylogenetic reconstructions cannot cope with these new datasets. Gubbins (Genealogies Unbiased By recomBinations In Nucleotide Sequences) is an algorithm that iteratively identifies loci containing elevated densities of base substitutions while concurrently constructing a phylogeny based on the putative point mutations outside of these regions. Simulations demonstrate the algorithm generates highly accurate reconstructions under realistic models of short-term bacterial evolution, and can be run in only a few hours on alignments of hundreds of bacterial genome sequences.</p>


<h2>Citing Gubbins</h2>

<p><em>Croucher N. J., Page A. J., Connor T. R., Delaney A. J., Keane J. A., Bentley S. D., Parkhill J., Harris S.R.<em> "Rapid phylogenetic analysis of large samples of recombinant bacterial whole genome sequences using Gubbins". (In submission), 2014.
</p>


<h2>Contact us</h2>
<p>If you need help or have any queries, please contact us using the details below. Our normal office hours are 8:30-17:00 (GMT) Monday to Friday.</p>
<ul>
  <li>Email <a href="mailto:gubbins@sanger.ac.uk">gubbins@sanger.ac.uk</a></li>
  <li>Phone Pathogen Informatics on +44 (0)1223 834244 extension 8736</li>
</ul>

<h3>User manual</h3>
<ul>
  <li>
    <a href="ftp://ftp.sanger.ac.uk/pub/project/pathogens/gubbins/GubbinsManualv0.6.docx">User manual</a>
  </li>
</ul>


<h2>
<a name="downloads" class="anchor" href="#downloads"><span class="octicon octicon-link"></span></a>Downloading and installing</h2>
<p>There are a number of options for installing Gubbins. We provide a few pre built downloads which vary depending on what system your running.</p>
<ul>
  <li><a href="#virtual-machine">Virtual machine</a> - For Windows, OSX, Linux</li>
  <li><a href="#trusty">Prebuilt Debian/Ubuntu package</a> - For Ubuntu Trusty</li>
  <li><a href="#linux_binary">Binary package</a> - For 64 bit Linux</li>
  <li><a href="#homebrew">Homebrew recipes</a> - OSX</li>
  <li><a href="#compiling">Compiling from source</a> - OSX, Linux</li>
</ul>

<h3>
<a name="virtual-machine" class="anchor" href="#virtual-machine"><span class="octicon octicon-link"></span></a>Virtual machine</h3>

<p>We have created virtual machine which has all of the software setup, along with the test datasets from the paper, based on <a href="http://environmentalomics.org/bio-linux/">Bio-Linux 8</a>. If you run Windows, then this is the only way to use Gubbins on your machine. You need to first install <a href="https://www.virtualbox.org/">VirtualBox</a>, then load the virtual machine, using the 'File -> Import Appliance' menu option. The root password is 'manager'.</p>

<ul>
<li><a href="ftp://ftp.sanger.ac.uk/pub/project/pathogens/gubbins/gubbins_0.6.ova">Virtual Machine with Gubbins installed</a></li>
</ul>



<h3>
<a name="trusty" class="anchor" href="#trusty"><span class="octicon octicon-link"></span></a>Ubuntu Trusty</h3>
<p>
  We have precompiled packages for Ubuntu (Trusty). You dont need to download anything, just copy and past the commands into a terminal. You need to first install DendroPy:
</p>
<ul>
  <li>wget http://pypi.python.org/packages/source/D/DendroPy/DendroPy-3.12.0.tar.gz
  </li>
  <li>tar xzvf DendroPy-3.12.0.tar.gz
  </li>
  <li>cd DendroPy-3.12.0
  </li>
  <li>sudo python setup.py install
  </li>
</ul>
<p>
  Then install gubbins
</p>
<ul>
  <li>sudo add-apt-repository ppa:ap13/gubbins
  </li>
  <li>sudo apt-get update
  </li>
  <li>sudo apt-get install fasttree raxml fastml2 gubbins
  </li>
</ul>


<h3>
<a name="compiling" class="anchor" href="#compiling"><span class="octicon octicon-link"></span></a>Precompiled packages</h3>
<p>We have prepared a binary package for amd 64 Linux. It was compiled on Ubuntu Trusty.  The dependancies will need to be manually installed following the instructions in the <a href="https://github.com/sanger-pathogens/gubbins/blob/master/INSTALL">INSTALL file</a>.</p> 
<ul>
  <li>
    <a href="ftp://ftp.sanger.ac.uk/pub/project/pathogens/gubbins/gubbins_0.6_amd64.tar.gz">Precompiled 64 bit Linux binary package</a>
  </li>
</ul>

<h3>
<a name="homebrew" class="anchor" href="#homebrew"><span class="octicon octicon-link"></span></a>Using Homebrew recipes</h3>
<p>If you are running OSX, you can use the homebrew package manager.</p>

<ul><li><a href="http://brew.sh/">Install homebrew from brew.sh</a></li>
  <li>Open a terminal and enter these commands</li>
  <li>brew tap homebrew/science</li>
  <li>brew install http://sanger-pathogens.github.io/gubbins/fastml.rb</li>
  <li>brew install http://sanger-pathogens.github.io/gubbins/gubbins.rb</li>  
</ul>


<h3>
<a name="virtual-machine" class="anchor" href="#compiling"><span class="octicon octicon-link"></span></a>Compiling from source</h3>
<p>
  Download the latest version of the <a href="https://github.com/sanger-pathogens/gubbins/tarball/master">Gubbins source code</a>, then follow the compile instructions in the <a href="https://github.com/sanger-pathogens/gubbins/blob/master/INSTALL">INSTALL file</a>.
</p>

      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Gubbins maintained by <a href="https://github.com/sanger-pathogens">sanger-pathogens</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
