<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Gubbins : Genealogies Unbiased By recomBinations In Nucleotide Sequences">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Gubbins - Genealogies Unbiased By recomBinations In Nucleotide Sequences</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/sanger-pathogens/gubbins">View source code on GitHub</a>

          <h1 id="project_title">Gubbins</h1>
          <h2 id="project_tagline">Genealogies Unbiased By recomBinations In Nucleotide Sequences</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/sanger-pathogens/gubbins/zipball/master">Download source code as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/sanger-pathogens/gubbins/tarball/master">Download source code as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a name="overview" class="anchor" href="#overview"><span class="octicon octicon-link"></span></a>Overview</h1>

<p>Since the introduction of high-throughput, second-generation DNA sequencing technologies, there has been an enormous increase in the size of datasets being used for estimating bacterial population phylodynamics. Although many phylogenetic techniques are scalable to hundreds of bacterial genomes, methods which have been used for mitigating the effect of mechanisms of horizontal sequence transfer on phylogenetic reconstructions cannot cope with these new datasets. Gubbins (Genealogies Unbiased By recomBinations In Nucleotide Sequences) is an algorithm that iteratively identifies loci containing elevated densities of base substitutions while concurrently constructing a phylogeny based on the putative point mutations outside of these regions. Simulations demonstrate the algorithm generates highly accurate reconstructions under realistic models of short-term bacterial evolution, and can be run in only a few hours on alignments of hundreds of bacterial genome sequences.</p>

<h2>The paper</h2>
The paper is available on Nucleic Acids Research (open access):
<a href="http://nar.oxfordjournals.org/content/early/2014/11/20/nar.gku1196.abstract">Rapid phylogenetic analysis of large samples of recombinant bacterial whole genome sequences using Gubbins</a>.

<h3>Citing Gubbins</h3>

<p><em>Croucher N. J., Page A. J., Connor T. R., Delaney A. J., Keane J. A., Bentley S. D., Parkhill J., Harris S.R.</em> "Rapid phylogenetic analysis of large samples of recombinant bacterial whole genome sequences using Gubbins". doi:10.1093/nar/gku1196, Nucleic Acids Research, 2014.
</p>

<h3>Data used in the paper</h3>
<ul>
  <li><a href="ftp://ftp.sanger.ac.uk/pub/project/pathogens/gubbins/PMEN1.aln.gz">PMEN1 dataset (12 Mbyte multi-FASTA alignment file)</a></li>
  <li><a href="ftp://ftp.sanger.ac.uk/pub/project/pathogens/gubbins/ST239.aln.gz">ST239 dataset (7 Mbyte multi-FASTA alignment file)</a></li>
</ul>

<h2>Contact us</h2>
<p>If you need help or have any queries, please contact us using the details below. Our normal office hours are 8:30-17:00 (GMT) Monday to Friday.</p>
<ul>
  <li>Email <a href="mailto:gubbins@sanger.ac.uk">gubbins@sanger.ac.uk</a></li>
  <li>Phone Pathogen Informatics on +44 (0)1223 834244 extension 8736</li>
</ul>

<h2>User manual</h2>
<ul>
  <li>
    <a href="ftp://ftp.sanger.ac.uk/pub/project/pathogens/gubbins/GubbinsManual_latest.pdf">User manual</a>
  </li>
</ul>

<h2>
<a name="downloads" class="anchor" href="#downloads"><span class="octicon octicon-link"></span></a>Downloading and installing</h2>
<p>There are a number of options for installing Gubbins. We provide a few pre built downloads which vary depending on what system your running.</p>
<ul>
  <li><a href="#virtual-machine">Virtual machine</a> - For Windows, OSX, Linux</li>
  <li><a href="#trusty">Prebuilt Debian/Ubuntu package</a> - For Ubuntu Trusty</li>
  <li><a href="#linux_binary">Binary package</a> - For 64 bit Linux</li>
  <li><a href="#homebrew">Homebrew recipes</a> - OSX</li>
  <li><a href="#compiling">Compiling from source</a> - OSX, Linux</li>
</ul>

<h3>
<a name="virtual-machine" class="anchor" href="#virtual-machine"><span class="octicon octicon-link"></span></a>Virtual machine</h3>

<p>We have created virtual machine which has all of the software setup, along with the test datasets from the paper, based on <a href="http://environmentalomics.org/bio-linux/">Bio-Linux 8</a>. If you run Windows, then this is the only way to use Gubbins on your machine. You need to first install <a href="https://www.virtualbox.org/">VirtualBox</a>, then load the virtual machine, using the 'File -> Import Appliance' menu option. The root password is 'manager'.</p>

<ul>
<li><a href="ftp://ftp.sanger.ac.uk/pub/pathogens/pathogens-vm/pathogens-vm.latest.ova">Virtual Machine with Gubbins installed</a></li>
</ul>



<h3>
<a name="trusty" class="anchor" href="#trusty"><span class="octicon octicon-link"></span></a>Ubuntu Trusty</h3>
<p>
  We have precompiled packages for Ubuntu (Trusty). You dont need to download anything, just copy and past the commands into a terminal. You need to first install DendroPy:
</p>
<ul>
  <li>wget http://pypi.python.org/packages/source/D/DendroPy/DendroPy-3.12.0.tar.gz
  </li>
  <li>tar xzvf DendroPy-3.12.0.tar.gz
  </li>
  <li>cd DendroPy-3.12.0
  </li>
  <li>sudo python setup.py install
  </li>
</ul>
<p>
  Then install gubbins
</p>
<ul>
  <li>sudo add-apt-repository ppa:ap13/gubbins
  </li>
  <li>sudo apt-get update
  </li>
  <li>sudo apt-get install fasttree raxml fastml2 gubbins
  </li>
</ul>


<h3>
<a name="linux_binary" class="anchor" href="#linux_binary"><span class="octicon octicon-link"></span></a>Precompiled packages</h3>
<p>We have prepared a binary package for amd 64 Linux. It was compiled on Ubuntu Trusty.  The dependancies will need to be manually installed following the instructions in the <a href="https://github.com/sanger-pathogens/gubbins/blob/master/INSTALL.md">INSTALL file</a>.</p> 
<ul>
  <li>
    <a href="ftp://ftp.sanger.ac.uk/pub/project/pathogens/gubbins/gubbins_latest_amd64.tar.gz">Precompiled 64 bit Linux binary package</a>
  </li>
</ul>

<h3>
<a name="homebrew" class="anchor" href="#homebrew"><span class="octicon octicon-link"></span></a>Using Homebrew recipes</h3>
<p>If you are running OSX, you can use the homebrew package manager.  These instructions are for OSX 10.9.4 (Mavericks). First of all you need to install homebrew, which involves also installing Xcode with command line tools. Then install the python modules, and finally the gubbins recipes.</p>

<ul>
  <li><a href="http://brew.sh/">Install homebrew from brew.sh</a></li>
  <li>Open a terminal and enter these commands</li>
  <li>curl https://bootstrap.pypa.io/ez_setup.py  | python - --user</li>
  <li>export PATH=~/Library/Python/2.7/bin:${PATH}</li>
  <li>easy_install pip</li>
  <li>pip install --user dendropy</li>
  <li>pip install --user biopython</li>
  <li>pip install --user reportlab</li>
  <li>brew tap homebrew/science</li>
  <li>brew install http://sanger-pathogens.github.io/gubbins/fastml.rb</li>
  <li>brew install http://sanger-pathogens.github.io/gubbins/gubbins.rb</li>  
</ul>


<h3>
<a name="virtual-machine" class="anchor" href="#compiling"><span class="octicon octicon-link"></span></a>Compiling from source</h3>
<p>
  Download the latest version of the <a href="https://github.com/sanger-pathogens/gubbins/tarball/master">Gubbins source code</a>, then follow the compile instructions in the <a href="https://github.com/sanger-pathogens/gubbins/blob/master/INSTALL">INSTALL file</a>.
</p>


<h1><a name="virtual-machine" class="anchor" href="#compiling"><span class="octicon octicon-link"></span></a>Running Gubbins</h1>
<p>To run Gubbins with default settings:</p>

    <ul><li>run_gubbins.py [FASTA alignment]</li></ul>
    
<p>Input options:</p>

    <ul><li>--outgroup, -o</li></ul>

<p>The name of a sequence in the alignment on which to root the tree</p>

    <ul><li>--starting_tree, -s</li></ul>

<p>A Newick-format starting tree on which to perform the first iteration analysis. The default is to compute a starting tree using RAxML</p>

    <ul><li>--filter_percentage -f</li></ul>

<p>Filter out taxa with more than this percentage of missing data. Default is 25%</p>
    
<p>Processing options:</p>

    <ul><li>--tree_builder, -t</li></ul>
    
<p>The algorithm to use in the construction of phylogenies in the analysis; can be ‘raxml’, to use RAxML, ‘fasttree’, to use Fasttree, or ‘hybrid’, to use Fasttree for the first iteration and RAxML in all subsequent iterations. Default is raxml</p>

    <ul><li>--iterations, -i</li></ul>
    
<p>The maximum number of iterations to perform; the algorithm will stop earlier than this if it converges on the same tree in two successive iterations. Default is 5.</p>

    <ul><li>--min_snps, -m</li></ul>
<p>The minimum number of base substitutions required to identify a recombination. Default is 3.</p>
 
    <ul><li>--converge_method, -z</li></ul>
<p>Convergence criteria. Criteria to use to know when to halt iterations (weighted_robinson_foulds, robinson_foulds, recombination). Default is 'weighted_robinson_foulds'.</p> 
    
<p>Output options:</p>

    <ul><li>--use_time_stamp, -u</li></ul>
    
<p>Include a time stamp in the name of output files to avoid overwriting previous runs on the same input file. Default is to not include a time stamp.</p>

    <ul><li>--prefix, -p</li></ul>
    
<p>Specifiy a prefix for output files. If none is provided it defaults to the name of the input FASTA alignment</p>

    <ul><li>--verbose, -v</li></ul>
    
<p>Print debugging messages. Default is off.</p>

    <ul><li>--no_cleanup, -n</li></ul>
    
<p>Do not remove files from intermediate iterations. This option will also keep other files created by RAxML, fastml and fasttree, which would otherwise be deleted. Default is to only keep files from the final iteration.</p>


<h2><a name="virtual-machine" class="anchor" href="#compiling"><span class="octicon octicon-link"></span></a>Output files</h2>
<h3><a name="virtual-machine" class="anchor" href="#compiling"><span class="octicon octicon-link"></span></a>Prefix</h3>
<p>If a prefix is not defined with the –prefix option, the default prefix of the output files is:
X.Y</p>

<p>where:</p>
<ul>
  <li>X = Prefix taken from the input fasta file</li>
<li>Y = Time stamp. NOTE: This will only be included in the output file prefix if the –u flag has been selected</li></ul>

<h3><a name="virtual-machine" class="anchor" href="#compiling"><span class="octicon octicon-link"></span></a>Output file suffices</h3>

<p>Recombination predictions in EMBL tab file format.</p>
<ul><li>.recombination_predictions.embl</li></ul>

<p>Recombination predictions in GFF3 format</p>
<ul><li>.recombination_predictions.gff	</li></ul>

<p>Base substitution reconstruction in EMBL tab format.</p>
<ul><li>.branch_base_reconstruction.embl	</li></ul>

<p>VCF file summarising the distribution of SNPs</p>
<ul><li>.summary_of_snp_distribution.vcf	</li></ul>

<p>Per branch reporting of the base substitutions inside and outside recombinations events.</p>
<ul><li>.per_branch_statistics.csv	</li></ul>

<p>FASTA format alignment of filtered polymorphic sites used to generate the phylogeny in the final iteration.</p>
<ul><li>.filtered_polymorphic_sites.fasta	</li></ul>

<p>Phylip format alignment of filtered polymorphic sites used to generate the phylogeny in the final iteration.</p>
<ul><li>.filtered_polymorphic_sites.phylip	</li></ul>

<p>Final phylogenetic tree in newick format.</p>
<ul><li>.final_tree.tre	</li></ul>


<h3>
<a name="virtual-machine" class="anchor" href="#compiling"><span class="octicon octicon-link"></span></a>Running on LSF</h3>
<p>The command to run the script using LSF with 8GB of RAM and 16 cores looks like:</p>
<ul><li>
bsub -o log -e err -M8000 -R "select[mem>8000] rusage[mem=8000]" -n16 -R "span[hosts=1]"  'run_gubbins.py --threads 16 my_alignment.aln'
</li></ul>

<p>If you are using the Sanger cluster (farm3), the avx option (-R "avx") will guarantee your analysis will run on the fastest processors using the AVX version of raxml:</p>
<ul><li>
bsub -R "avx" -o log -e err -M8000 -R "select[mem>8000] rusage[mem=8000]" -n16 -R "span[hosts=1]"  'run_gubbins.py --threads 16 my_alignment.aln'
</li></ul>


      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Gubbins maintained by <a href="https://github.com/sanger-pathogens">sanger-pathogens</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
